src/parse.ts

```diff
@@ -43,23 +43,23 @@ export function parseEnv(options: IGetEnvOptions) {
   }
 }
 
-export function parse<T extends IEnvDefinition>(definition: T) {
-  // TODO: replace any with result type of T[k] when typescript supports it
-  //       see https://github.com/Microsoft/TypeScript/issues/6606
-  const parsed: {[k in keyof T]?: any} = {};
+export function parse<S extends object>(definition: IEnvDefinition<S>) {
+  const parsed: {[P in keyof S]?: S[P]} = {};
 
-  Object.keys(definition).forEach(key => {
-    const value = definition[key];
+  for (const key in definition) {
```

Comment:
Unfortunately, that only works witha `for .... in` loop, which requires a call to `hasOwnProperty`.

Reply:
For `Object.keys(definition).forEach(key => {...})` TypeScript infers the type of `key` to be `string`, because the type definition of `Object.keys` is `(o: {}) => string[]`. This turns out to be too weak for `S[typeof key]` in line 51 to be well-defined.

For `for (const key in definition) {...}` TypeScript infers the type of `key` to be `Extract<keyof S, string>` (a TypeScript 2.8 concept, which is basically just `keyof S`). This is correct and strong enough.

Reply:
got it, thanks for the explanation! :)